Source/NSGridView.m

```diff
@@ -0,0 +1,1232 @@
+
+  FOR_IN(NSGridCell*, c, _cells)
+    {
+      if ([c column] == col)
+        {          
+          [result addObject: c];
+        }
+    }
+  END_FOR_IN(_cells);
+  
+  return result;
+}
+
+- (NSArray *) _viewsForColumnAtIndex: (NSUInteger)columnIndex
+{
+  NSArray *cells = [self _cellsForColumnAtIndex: columnIndex];
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity: [_rows count]];
+  
+  FOR_IN(NSGridCell*, c, cells)
+    {
+      NSView *v = [c contentView];
+      if (c != nil)
+        {
+          [result addObject: v];
+        }
+      else
+        {
+          [result addObject: [NSGridCell emptyContentView]];
+        }
+    }
+  END_FOR_IN(cells);
+
+  return result;
+}
+@end
+
+
+// Private methods...
+@implementation NSGridCell (Private)
+- (void) _setOwningRow: (NSGridRow *)r
+{
+  _owningRow = r;
+}
+
+- (void) _setOwningColumn: (NSGridColumn *)c
+{
+  _owningColumn = c;
+}
+@end
+
+@implementation NSGridView
+
++ (void) initialize
+{
+  if (self == [NSGridView class])
+    {
+      [self setVersion: 1];
+    }
+}
+
+- (void) _refreshCells
+{
+  if ([self isHidden])
+    {
+      return;
+    }
+  else
+    {
+      NSUInteger i = 0;
+      NSUInteger num_col = [self numberOfColumns];
+      NSRect f = [self frame];
+      CGFloat current_x = 0.0, current_y = f.size.height;
+      NSRect p = [self _prototypeFrame];
+      
+      // Format the grid...
+      FOR_IN(NSGridCell*, c, _cells)
+        {
+          NSView *v = [c contentView];
+          NSUInteger ri = 0, ci = 0;
+          
+          // Get row and column index...
+          ci = i % num_col;
+          ri = i / num_col;
+          
+          // Get the row and col...
+          NSGridRow *row = [self rowAtIndex: ri];
+          NSGridColumn *col = [self columnAtIndex: ci];
+          NSRect rect = NSZeroRect;
+          
+          if (v != nil)
+            {
+              rect = [v frame];
+            }
+          
+          // Do the math for the frame...
+          // NOTE: I am not sure why this is needed, the row and column heights are coming in via the nib
+          // as VERY small numbers (1.175... e-38) (see PR for this work, I have attached the XML nib there)
+          // so it is necessary to figure out the width of each column so that things can be properly spaced.
+          if (ci == 0)
+            {
+              current_y -= p.size.height; // rect.size.height;
+              current_x = 0.0;
+            }
+          
+          current_y -= [c yPlacement] - [row topPadding];
+          current_x += [c xPlacement] + [col leadingPadding] + [col width];
+          
+          if (v != nil)
+            {
+              rect = p;
+              rect.origin.x = current_x;
+              rect.origin.y = current_y;
+              [v setFrame: rect];
+              [self addSubview: v];
+            }
+          
+          current_x += [col trailingPadding] + [col width] + p.size.width + _columnSpacing;
```

Comment:
Having both [col width]  and p.size.width here seems to be too much. Did you ever run this method and had some values output from it?

Reply:
I believe p.size.width is the correct thing to use here.